lib/Middleware.js

```diff
@@ -0,0 +1,223 @@
+/*!
+ * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+// FIXME: use did-io
+const axios = require('axios');
+const didio = {get: () => {}};
+const httpSignatureHeader = require('http-signature-header');
+const httpSignatureCrypto = require('http-signature-crypto');
+const jsonld = require('jsonld');
+const url = require('url');
+const {HttpSignatureError} = httpSignatureHeader;
+const {Strategy} = require('passport-strategy');
+
+// source: https://stackoverflow.com/a/3143231
+const dateRegEx = /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/;
+
+module.exports = class Middleware extends Strategy {
+  constructor({name}) {
+    super();
+    this.name = name || 'http-signature-strategy';
+    this._helpers = {};
+    this.dereferenceUrlScheme = {did: true, https: true};
+  }
+
+  async authenticate(req) {
+    const self = this;
+
+    // parse the req
+    let parsed;
+    try {
+      parsed = httpSignatureHeader.parseRequest(req);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    const publicKey = {id: parsed.keyId};
+
+    const getKey = self.use('getKey');
+
+    if(!getKey) {
+      return self.error(new HttpSignatureError(
+        'A `getKey` function must be defined.', 'InvalidStateError'));
+    }
+
+    let keyDoc;
+    try {
+      keyDoc = await getKey(publicKey);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    try {
+      await self._verifySignature({keyDoc, parsed});
+    } catch(err) {
+      return self.error(err);
+    }
+
+    const validateRequest = self.use('validateRequest');
+    if(!validateRequest) {
+      return self.error(new HttpSignatureError(
+        'A `validateRequest` function must be defined.', 'InvalidStateError'));
+    }
+    try {
+      await validateRequest(req);
+    } catch(err) {
+      return self.error(err);
+    }
+
+    const getUser = self.use('getUser') || self.getUser.bind(this);
+
+    // getUser must throw error if the user/identity is not found
+    let user;
+    try {
+      user = await getUser({keyDoc});
```

Comment:
I think we actually also need to pass `req` to `getUser`. For object capability use cases we will need to pull capabilities out of the request headers and add them to `actor` which we attach to `user`. It may turn out that `capability` becomes a parameter in the signature ... or may become just another http header to sign (we don't know yet). That may mean we could pass the capabilities in too to `getUser` ... or we could just have `getUser` call `parse` on the signature header and pull out the capabilities there. So I think either way it would be useful to pass `req` to `getUser` ... we don't know what other information (probably from headers) may be used to help construct the user.